py/objstr: bytes(bytes_obj) is bytes_obj

iyassou opened the pull request:
Calling the `bytes` constructor on a `bytes` object returns the original `bytes` object. This addresses #6320 and has been tested on an ESP32, and everything seems to be working fine:

```python
MicroPython v1.13-dirty on 2020-09-18; ESP32 module with ESP32
Type "help()" for more information.
>>> a = b"Original"
>>> b = a
>>> c = bytes(a)
>>> type(a)
<class 'bytes'>
>>> a is b
True
>>> a is c
True
>>> id(a)
1073642736
>>> id(b)
1073642736
>>> id(c)
1073642736
```

dpgeorge commented:
Thanks for the patch.  I guess it makes sense, if only from an efficiency point of view to not unnecessarily allocate a new object.

Note: CI failure is due to code-size increase of 12 to 16 bytes, and trailing whitespace on a blank line.

dpgeorge commented:
> CI failure is due to code-size increase of 12 to 16 bytes

That bit can't be helped, so it's ok to fail.

iyassou commented:
> 
> 
> > CI failure is due to code-size increase of 12 to 16 bytes
> 
> That bit can't be helped, so it's ok to fail.

Noted. The trailing whitespace on the blank line has been removed now.

peterhinch commented:
>  I guess it makes sense, if only from an efficiency point of view to not unnecessarily allocate a new object.

Also to match CPython:
```python
a = b'rats'
b = bytes(a)
a is b  #  `True` in CPython but `False` in MP.
```
The behaviour of `str` is OK in this regard.

jimmo commented:
I think this is a sensible optimisation, but we actually somewhat rely on this behaviour in modbluetooth if the user wants to make a copy of the advertising data or address.   modbluetooth passes a pre-allocated mp_type_bytes instance with the current event's data to the IRQ handler, and the user can use `bytes(addr)` to copy it.

We also suggest that in the docs:

```If your program wants to use this data outside of the handler, then it must copy them first, e.g. by using ``bytes(addr)`` or ``bluetooth.UUID(uuid)``.```

I guess we could make modbluetooth use a bytearray instead of bytes?

iyassou commented:
Although I recently used `ubluetooth`, I had completely forgotten about this.

> I guess we could make modbluetooth use a bytearray instead of bytes?

That does work around this optimisation. The following example application is working fine:

- **`ble_bytearray_test.py`**

```python
import ubluetooth as bluetooth
from ble_advertising import advertising_payload
from micropython import const

_IRQ_SCAN_RESULT = const(5)

class Test:
    def __init__(self):
        name = 'Working 👍'.encode('utf-8') # Unicode names are a pleasant sight!
        ble = bluetooth.BLE()
        ble.active(True)
        ble.config(gap_name=name)
        ble.irq(handler=self.irq)
        self.payload = advertising_payload(name=name)
        ble.gap_advertise(500000, adv_data=self.payload)
        ble.gap_scan(0)

    def irq(self, event, data):
        if event == _IRQ_SCAN_RESULT:
            _, addr, _, _, adv_data = data
            addr_ = bytearray(addr)
            adv_data_ = bytearray(adv_data)
            print('addr:', addr_)
            print('adv_data:', adv_data_)
            for i in range(len(addr_)):
                addr_[i] = 5
            for i in range(len(adv_data_)):
                adv_data_[i] = 121
            print('addr modded:', addr_)
            print('addr original:', addr)
            print('adv_data modded:', adv_data_)
            print('adv_data original:', adv_data)

```

- **`main.py`**

```python
from ble_bytearray_test import Test
t = Test()
```

- **REPL**

```python
MPY: soft reboot
MicroPython v1.13-dirty on 2020-09-18; ESP32 module with ESP32
Type "help()" for more information.
>>> addr: bytearray(b'8\x01\x95\x8eB1')
adv_data: bytearray(b'\x1b\xffu\x00B\x04\x01\x80 8\x01\x95\x8eB1:\x01\x95\x8eB0\x01\x00\x00\x00\x00\x00\x00')
addr modded: bytearray(b'\x05\x05\x05\x05\x05\x05')
addr original: b'8\x01\x95\x8eB1'
adv_data modded: bytearray(b'yyyyyyyyyyyyyyyyyyyyyyyyyyyy')
adv_data original: b'\x1b\xffu\x00B\x04\x01\x80 8\x01\x95\x8eB1:\x01\x95\x8eB0\x01\x00\x00\x00\x00\x00\x00'
```

I wonder how many other applications rely on the `bytes` constructor's behaviour.

dpgeorge commented:
> CI failure is due to code-size increase of 12 to 16 bytes

That bit can't be helped, so it's ok to fail.

> but we actually somewhat rely on this behaviour in modbluetooth if the user wants to make a copy of the advertising data or address. modbluetooth passes a pre-allocated mp_type_bytes instance with the current event's data to the IRQ handler, and the user can use `bytes(addr)` to copy it.

Yes, very good point.  I think using a `bytearray` is not good because the user can try to resize it.  Instead a `memoryview` would be better (can't be resized, can even be made read-only).  A `memoryview` actually describes the concept quite well, and might help users remember to copy it when they try and print it out.

dpgeorge commented:
@iyassou I see you use a no-reply github email address for this commit.  Are you able to use a real one?  Please see the first section of `CODECONVENTIONS.md`.

iyassou committed:
py/objstr: bytes(bytes_obj) is bytes_obj

Calling the bytes constructor on a bytes object returns the original bytes object.

Signed-off-by: Iyassou Shimels <s.iyassou@gmail.com>

iyassou commented:
> > but we actually somewhat rely on this behaviour in modbluetooth if the user wants to make a copy of the advertising data or address. modbluetooth passes a pre-allocated mp_type_bytes instance with the current event's data to the IRQ handler, and the user can use `bytes(addr)` to copy it.
> 
> Yes, very good point. I think using a `bytearray` is not good because the user can try to resize it. Instead a `memoryview` would be better (can't be resized, can even be made read-only). A `memoryview` actually describes the concept quite well, and might help users remember to copy it when they try and print it out.

It appears I misundertood what was suggested. Indeed a `memoryview` would be just what was needed and more representative of the underlying setup.

iyassou commented:
> 
> 
> @iyassou I see you use a no-reply github email address for this commit. Are you able to use a real one? Please see the first section of `CODECONVENTIONS.md`.

I've added an email address I'm reachable at and tidied up my pull request.

dpgeorge commented:
Thanks for updating, rebased and merged in ca017841d65b5f9da5eb4e6e3ca66b009bc54fc4
